text/0000-api-evolution.md

```diff
@@ -0,0 +1,748 @@
+
+#### Minor change: adding or removing private fields when at least one already exists.
+
+No existing code could be relying on struct literals for the struct, nor on
+exhaustively matching its contents, and client code will likewise be oblivious
+to the addition of further private fields.
+
+For tuple structs, this is only a minor change if furthermore *all* fields are
+currently private. (Tuple structs with mixtures of public and private fields are
+bad practice in any case.)
+
+#### Minor change: going from a tuple struct with all private fields (with at least one field) to a normal struct, or vice versa.
+
+This is technically a breaking change:
+
+```rust
+// in some other module:
+pub struct Foo(SomeType);
+
+// in downstream code
+let Foo(_) = foo;
+```
+
+Changing `Foo` to a normal struct can break code that matches on it -- but there
+is never any real reason to match on it in that circumstance, since you cannot
+extract any fields or learn anything of interest about the struct.
+
+### Enums
+
+See "[Signatures in type definitions](#signatures-in-type-definitions)" for some
+general remarks about changes to the actual types in an `enum` definition.
+
+#### Major change: adding new variants.
+
+Exhaustiveness checking means that a `match` that explicitly checks all the
+variants for an `enum` will break if a new variant is added. It is not currently
+possible to defend against this breakage in advance.
+
+A [postponed RFC](https://github.com/rust-lang/rfcs/pull/757) discusses a
+language feature that allows an enum to be marked as "extensible", which
+modifies the way that exhaustiveness checking is done and would make it possible
+to extend the enum without breakage.
+
+#### Major change: adding new fields to a variant.
+
+If the enum is public, so is the full contents of all of its variants. As per
+the rules for structs, this means it is not allowed to add any new fields (which
+will automatically be public).
+
+If you wish to allow for this kind of extensibility, consider introducing a new,
+explicit struct for the variant up front.
+
+### Traits
+
+#### Major change: adding a non-defaulted item.
+
+Adding any item without a default will immediately break all trait implementations.
+
+It's possible that in the future we will allow some kind of
+"[sealing](#sealed-traits)" to say that a trait can only be used as a bound, not
+to provide new implementations; such a trait *would* allow arbitrary items to be
+added.
+
+#### Major change: any non-trivial change to item signatures.
+
+Because traits have both implementors and consumers, any change to the signature
+of e.g. a method will affect at least one of the two parties. So, for example,
+abstracting a concrete method to use generics instead might work fine for
+clients of the trait, but would break existing implementors. (Note, as above,
+the potential for "sealed" traits to alter this dynamic.)
+
+#### Minor change: adding a defaulted item.
```

Comment:
Adding a defaulted method to a trait that has any associated types with a default should probably be considered a major change.

In particular, due to [Default handling in RFC 195](https://github.com/rust-lang/rfcs/blob/master/text/0195-associated-items.md#defaults), if I have an impl that overrides a defaulted associated type, then I will be obligated to now implement the associated method that the upstream code added, regardless of the fact that it carried a default.  (At least, that's my memory of how that rule is to be interpreted.)